File: dart/lib/src/parameter_type_matcher.dart
import 'package:cucumber_expressions/src/parameter_type.dart';

final RegExp _wordBoundary = RegExp(r'\p{Z}|\p{P}|\p{S}', unicode: true);

/// Matches a single parameter type's regular expression against text, used by
/// the expression generator to find candidate parameters.
class ParameterTypeMatcher {
  /// Creates a matcher for [parameterType] using [_regexpString] against
  /// [_text], starting at [_matchPosition].
  ParameterTypeMatcher(
    this.parameterType,
    this._regexpString,
    this._text, [
    this._matchPosition = 0,
  ]) : _match = RegExp('($_regexpString)').firstMatch(
          _text.substring(_matchPosition),
        );

  /// The parameter type this matcher represents.
  final ParameterType<Object?> parameterType;
  final String _regexpString;
  final String _text;
  final int _matchPosition;
  final RegExpMatch? _match;

  /// Returns a matcher advanced to the next match at or after
  /// [newMatchPosition].
  ParameterTypeMatcher advanceTo(int newMatchPosition) {
    for (var advancedPos = newMatchPosition;
        advancedPos < _text.length;
        advancedPos++) {
      final matcher = ParameterTypeMatcher(
        parameterType,
        _regexpString,
        _text,
        advancedPos,
      );
      if (matcher.find) {
        return matcher;
      }
    }
    return ParameterTypeMatcher(
      parameterType,
      _regexpString,
      _text,
      _text.length,
    );
  }

  /// Whether a non-empty, full-word match was found.
  bool get find => _match != null && group != '' && fullWord;

  /// The start index of the match within the text.
  int get start {
    final match = _match;
    if (match == null) {
      throw StateError('No match');
    }
    return _matchPosition + match.start;
  }

  /// Whether the match starts and ends on word boundaries.
  bool get fullWord => matchStartWord && matchEndWord;

  /// Whether the match starts on a word boundary.
  bool get matchStartWord =>
      start == 0 || _wordBoundary.hasMatch(_text[start - 1]);

  /// Whether the match ends on a word boundary.
  bool get matchEndWord {
    final nextCharacterIndex = start + group.length;
    return nextCharacterIndex == _text.length ||
        _wordBoundary.hasMatch(_text[nextCharacterIndex]);
  }

  /// The matched substring.
  String get group {
    final match = _match;
    if (match == null) {
      throw StateError('No match');
    }
    return match.group(0)!;
  }

  /// Orders matchers by start position, then by descending match length.
  static int compare(ParameterTypeMatcher a, ParameterTypeMatcher b) {
    final posComparison = a.start - b.start;
    if (posComparison != 0) {
      return posComparison;
    }
    final lengthComparison = b.group.length - a.group.length;
    if (lengthComparison != 0) {
      return lengthComparison;
    }
    return 0;
  }
}

File: elixir/lib/cucumber/cucumber_expressions/parameter_type_matcher.ex
defmodule Cucumber.CucumberExpressions.ParameterTypeMatcher do
  @moduledoc false

  # Finds where a parameter type's regexp matches inside a step text, used by
  # the generator. Positions are byte offsets into the text; `advance_to/2`
  # only probes codepoint boundaries.

  defstruct [:parameter_type, :regexp, :text, :offsets, :match]

  @word_boundary ~r/\p{Z}|\p{P}|\p{S}/u

  def new(parameter_type, regexp, text, match_position) do
    %__MODULE__{
      parameter_type: parameter_type,
      regexp: regexp,
      text: text,
      offsets: codepoint_offsets(text)
    }
    |> probe(match_position)
  end

  # Re-probes an existing matcher at `match_position`, reusing its cached
  # codepoint offsets rather than re-scanning the text.
  defp probe(%__MODULE__{regexp: regexp, text: text} = matcher, match_position) do
    match =
      with true <- match_position <= byte_size(text),
           [{match_start, match_length}, {group_start, group_length} | _] <-
             Regex.run(regexp, text, offset: match_position, return: :index) do
        %{
          start: match_start,
          length: match_length,
          group: binary_part(text, group_start, group_length)
        }
      else
        _ -> nil
      end

    %{matcher | match: match}
  end

  def find?(%__MODULE__{match: nil}), do: false
  def find?(%__MODULE__{match: %{group: group}}), do: group != ""

  @doc """
  Returns the first matcher at or after `new_match_position` whose match is a
  full word,
  or a matcher positioned at the end of the text.
  """
  def advance_to(%__MODULE__{text: text, offsets: offsets} = matcher, new_match_position) do
    offsets
    |> Enum.drop_while(&(&1 < new_match_position))
    |> Enum.find_value(fn offset ->
      candidate = probe(matcher, offset)
      if find?(candidate) and full_word?(candidate), do: candidate
    end)
    |> Kernel.||(probe(matcher, byte_size(text)))
  end

  def sorts_before?(%__MODULE__{match: a}, %__MODULE__{match: b}) do
    if a.start != b.start do
      a.start < b.start
    else
      String.length(a.group) >= String.length(b.group)
    end
  end

  def same_rank?(%__MODULE__{match: a}, %__MODULE__{match: b}) do
    a.start == b.start and String.length(a.group) == String.length(b.group)
  end

  defp full_word?(%__MODULE__{text: text, match: %{start: start, length: length}}) do
    boundary_before?(text, start) and boundary_after?(text, start + length)
  end

  defp boundary_before?(_text, 0), do: true

  defp boundary_before?(text, match_start) do
    boundary?(String.last(binary_part(text, 0, match_start)))
  end

  defp boundary_after?(text, match_end) when match_end == byte_size(text), do: true

  defp boundary_after?(text, match_end) do
    boundary?(String.first(binary_part(text, match_end, byte_size(text) - match_end)))
  end

  defp boundary?(char), do: Regex.match?(@word_boundary, char)

  defp codepoint_offsets(text) do
    {offsets, _} =
      text
      |> String.codepoints()
      |> Enum.map_reduce(0, fn codepoint, offset ->
        {offset, offset + byte_size(codepoint)}
      end)

    offsets
  end
end
